File: src/Zametek.Maths.Graphs.Primitives/Entities/Node.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Zametek.Maths.Graphs
{
    /// <summary>
    /// A directed-graph node carrying a content payload (an event in arrow graphs; an activity in vertex graphs). Equality is by ID only.
    /// </summary>
    public class Node<T, TContent>
        : IHaveId<T>, IHaveContent<TContent>, IEquatable<Node<T, TContent>>, ICloneObject
        where T : struct, IComparable<T>, IEquatable<T>
        where TContent : IHaveId<T>, ICloneObject
    {
        #region Fields

        private readonly HashSet<T> m_IncomingEdges;
        private readonly HashSet<T> m_OutgoingEdges;

        private const int c_HashFactorOne = 17;
        private const int c_HashFactorTwo = 23;

        #endregion

        #region Ctors

        /// <summary>
        /// Creates a normal node carrying the given content.
        /// </summary>
        public Node(TContent content)
            : this(NodeType.Normal, content)
        {
        }

        /// <summary>
        /// Creates a node of the given type carrying the given content.
        /// </summary>
        public Node(NodeType nodetype, TContent content)
        {
            if (content == null)
            {
                throw new ArgumentNullException(nameof(content));
            }
            NodeType = nodetype;
            Content = content;
            m_IncomingEdges = new HashSet<T>();
            m_OutgoingEdges = new HashSet<T>();
        }

        #endregion

        #region Properties

        /// <inheritdoc/>
        public T Id => Content.Id;

        /// <summary>
        /// The position of the node within the graph.
        /// </summary>
        public NodeType NodeType
        {
            get;
            private set;
        }

        /// <inheritdoc/>
        public TContent Content
        {
            get;
        }

        /// <summary>
        /// The IDs of the edges pointing into this node (invalid for Start and Isolated nodes).
        /// </summary>
        public HashSet<T> IncomingEdges
        {
            get
            {
                if (NodeType == NodeType.Start || NodeType == NodeType.Isolated)
                {
                    throw new InvalidOperationException(Properties.Resources.Message_CannotRequestIncomingEdgesOfStartOrIsolatedNode);
                }
                return m_IncomingEdges;
            }
        }

        /// <summary>
        /// The IDs of the edges leaving this node (invalid for End and Isolated nodes).
        /// </summary>
        public HashSet<T> OutgoingEdges
        {
            get
            {
                if (NodeType == NodeType.End || NodeType == NodeType.Isolated)
                {
                    throw new InvalidOperationException(Properties.Resources.Message_CannotRequestOutgoingEdgesOfEndOrIsolatedNode);
                }
                return m_OutgoingEdges;
            }
        }

        #endregion

        #region Public Methods

        /// <summary>
        /// Changes the position classification of the node.
        /// </summary>
        public void SetNodeType(NodeType nodeType)
        {
            NodeType = nodeType;
        }

        #endregion

        #region Overrides

        /// <inheritdoc/>
        public override bool Equals(object? obj)
        {
            return Equals(obj as Node<T, TContent>);
        }

        /// <inheritdoc/>
        public override int GetHashCode()
        {
            unchecked
            {
                int hash = c_HashFactorOne;
                hash = hash * c_HashFactorTwo + Id.GetHashCode();
                hash = hash * c_HashFactorTwo + NodeType.GetHashCode();
                return hash;
            }
        }

        #endregion

        #region IEquatable

        /// <inheritdoc/>
        public bool Equals(Node<T, TContent>? other)
        {
            if (other is null)
            {
                return false;
            }
            return Id.Equals(other.Id)
                    && NodeType == other.NodeType
                    && m_IncomingEdges.OrderBy(x => x).SequenceEqual(other.m_IncomingEdges.OrderBy(x => x))
                    && m_OutgoingEdges.OrderBy(x => x).SequenceEqual(other.m_OutgoingEdges.OrderBy(x => x));
        }

        #endregion

        #region ICloneObject

        /// <inheritdoc/>
        public virtual object CloneObject()
        {
            var output = new Node<T, TContent>(NodeType, (TContent)Content.CloneObject());
            foreach (T edgeId in m_IncomingEdges)
            {
                output.m_IncomingEdges.Add(edgeId);
            }
            foreach (T edgeId in m_OutgoingEdges)
            {
                output.m_OutgoingEdges.Add(edgeId);
            }
            return output;
        }

        #endregion
    }
}

File: rust/primitives/src/node.rs
use crate::edge::HasId;
use crate::enums::NodeType;
use crate::error::GraphError;
use crate::insertion_order_set::InsertionOrderSet;
use crate::key::Key;

pub const MSG_CANNOT_REQUEST_INCOMING_EDGES: &str =
    "Cannot request Incoming Edges of a Start or Isolated Node";
pub const MSG_CANNOT_REQUEST_OUTGOING_EDGES: &str =
    "Cannot request Outgoing Edges of an End or Isolated Node";

/// A directed-graph node carrying a content payload (an event in arrow graphs;
/// an activity in vertex graphs) - the counterpart of the C# `Node<T, TContent>`.
///
/// The C# `IncomingEdges`/`OutgoingEdges` properties throw for node types that
/// cannot have them; the checked accessors [`Node::incoming_edges`] and
/// [`Node::outgoing_edges`] reproduce that contract. The raw sets are also
/// exposed (`incoming`/`outgoing`) for algorithm code that has already
/// established the node type, as the C# internals do.
///
/// The edge sets iterate in insertion order, which the golden tests encode, but
/// remove in O(1) - see [`InsertionOrderSet`] for why that matters here.
#[derive(Debug, Clone)]
pub struct Node<K: Key, C> {
    node_type: NodeType,
    pub content: C,
    pub incoming: InsertionOrderSet<K>,
    pub outgoing: InsertionOrderSet<K>,
}

impl<K: Key, C: HasId<K>> Node<K, C> {
    /// Creates a normal node carrying the given content.
    pub fn new(content: C) -> Self {
        Self::with_type(NodeType::Normal, content)
    }

    /// Creates a node of the given type carrying the given content.
    pub fn with_type(node_type: NodeType, content: C) -> Self {
        Self {
            node_type,
            content,
            incoming: InsertionOrderSet::new(),
            outgoing: InsertionOrderSet::new(),
        }
    }

    pub fn id(&self) -> K {
        self.content.id()
    }

    /// The position of the node within the graph.
    pub fn node_type(&self) -> NodeType {
        self.node_type
    }

    /// Changes the position classification of the node.
    pub fn set_node_type(&mut self, node_type: NodeType) {
        self.node_type = node_type;
    }

    /// The IDs of the edges pointing into this node (invalid for Start and Isolated nodes).
    pub fn incoming_edges(&self) -> Result<&InsertionOrderSet<K>, GraphError> {
        if matches!(self.node_type, NodeType::Start | NodeType::Isolated) {
            return Err(GraphError::new(MSG_CANNOT_REQUEST_INCOMING_EDGES));
        }
        Ok(&self.incoming)
    }

    /// The IDs of the edges leaving this node (invalid for End and Isolated nodes).
    pub fn outgoing_edges(&self) -> Result<&InsertionOrderSet<K>, GraphError> {
        if matches!(self.node_type, NodeType::End | NodeType::Isolated) {
            return Err(GraphError::new(MSG_CANNOT_REQUEST_OUTGOING_EDGES));
        }
        Ok(&self.outgoing)
    }
}

impl<K: Key, C: HasId<K>> PartialEq for Node<K, C> {
    /// Equality by ID, node type and edge sets (order-insensitive), as in C#.
    fn eq(&self, other: &Self) -> bool {
        if self.id() != other.id() || self.node_type != other.node_type {
            return false;
        }
        let mut a: Vec<K> = self.incoming.iter().copied().collect();
        let mut b: Vec<K> = other.incoming.iter().copied().collect();
        a.sort();
        b.sort();
        if a != b {
            return false;
        }
        let mut a: Vec<K> = self.outgoing.iter().copied().collect();
        let mut b: Vec<K> = other.outgoing.iter().copied().collect();
        a.sort();
        b.sort();
        a == b
    }
}

impl<K: Key, C: HasId<K>> Eq for Node<K, C> {}
